tests/testthat/test-calculate_tdm.R

```diff
@@ -0,0 +1,160 @@
+test_that("with bad `data` errors with informative message", {
+  expect_error(calculate_tdm("bad", 2020), "data.frame.*not.*TRUE")
+  expect_error(calculate_tdm(fake_tdm_data(), "bad"), "numeric.*not.*TRUE")
+  expect_error(calculate_tdm(fake_tdm_data(), 2020, t1 = "bad"), "numeric.*not.*TRUE")
+  expect_error(calculate_tdm(fake_tdm_data(), 2020, t2 = "bad"), "numeric.*not.*TRUE")
+  expect_error(calculate_tdm(fake_tdm_data(), 2020, additional_groups = TRUE), "character.*not.*TRUE")
+})
+
+test_that("outputs the expected tibble", {
+  data <- fake_tdm_data(
+    year = c(2020, 2025, 2030),
+    scen_alloc_wt_tech_prod = 1:3
+  )
+
+  out <- calculate_tdm(data, 2020)
+  expect_snapshot(out)
+})
+
+test_that("outputs 0 for `tdm_tech` and `tdm_sec`, if `scen_alloc_wt_tech_prod
+          doesn't change over 10 year period`", {
+  data <- fake_tdm_data(year = c(2020, 2025, 2030))
+  out <- calculate_tdm(data, 2020)
+  expect_true(out$tdm_tech == 0)
+  expect_true(out$tdm_sec == 0)
+})
+
+test_that("outputs is ungrouped", {
+  data <- fake_tdm_data(year = c(2020, 2025, 2030))
+  out <- calculate_tdm(data, 2020)
+  expect_false(is_grouped_df(out))
+})
+
+test_that("joins quietly", {
+  data <- fake_tdm_data(year = c(2020, 2025, 2030))
+  expect_no_message(calculate_tdm(data, 2020))
+})
+
+test_that("with data lacking crucial columns errors with informative message", {
+  expect_error_missing_names <- function(name) {
+    pacta_results <- fake_tdm_data(year = c(2020, 2025, 2030))
+    bad <- rename(pacta_results, bad = all_of(name))
+    expect_error(calculate_tdm(bad, 2020), class = "missing_names")
+  }
+
+  expect_error_missing_names("allocation")
+  expect_error_missing_names("ald_sector")
+  expect_error_missing_names("technology")
+  expect_error_missing_names("year")
+  expect_error_missing_names("plan_carsten")
+  expect_error_missing_names("scen_alloc_wt_tech_prod")
+  expect_error_missing_names("plan_alloc_wt_tech_prod")
+})
+
+test_that("if only `allocation` is 'ownership_weight' outputs a 0-row tibble", {
+  fake_tdm_data(allocation = "ownership_weight", year = c(2020, 2025, 2030)) %>%
+    calculate_tdm(2020) %>%
+    expect_s3_class("tbl") %>%
+    expect_warning(class = "has_zero_rows")
+})
+
+test_that("errors if data has multiple values per year", {
+  fake_tdm_data(
+    year = c(2020, 2020, 2025, 2030),
+    plan_alloc_wt_tech_prod = c(1, 1, 3, 4),
+    scen_alloc_wt_tech_prod = c(1, 2, 3, 4),
+    plan_carsten = c(0.5, 0.3, 0.25, 0.2),
+  ) %>%
+    calculate_tdm(2020) %>%
+    expect_error(class = "multiple_values_per_year")
```

Comment:
Consider `expect_snapshot_error()` to confirm the message prints with the format you want.